Add GPT fine-tunning and Vicuna deployment docs and codes (#11)
XinShuYang/antrea: ai/docs/Vicuna Model Deployment and Training Documentation.md

        "id": "1",
        "conversations": [
            {
                "from": "human",
                "value": "What is antctl and what are its different operating modes"
            },
            {
                "from": "gpt",
                "value": "Antctl is the command-line tool for Antrea, a significant component in the Kubernetes networking framework. It operates in three distinct modes: controller mode, agent mode, and standalone mode. In controller mode, it can connect to the Antrea Controller when run out-of-cluster or within the Controller Pod, enabling queries about network policies. In agent mode, when executed within an Antrea Agent Pod, it interacts with the local agent to provide information specific to that Agent, such as the set of computed Network Policies. The standalone mode is not explicitly described in the provided excerpt of the article."
            }
        ]
    },
```
An example is like "./antrea/ai/vicuna/dataset/dummy_conversation.json". After the model and dataset being set ready, you should change to the "./antrea/ai/vicuna/train_full" directory for full parameter fine-tuning, or "./antrea/ai/vicuna/train_qlora" for QLoRA fine-tuning.

> ### 2.Build container for training

Reviewer: To avoid confusion with pretraining, using fine-tuning is more accurate and helps prevent misunderstandings. https://github.com/lm-sys/FastChat#fine-tuning

Reviewer: Please also update the following content to make the description more accurate.

```plain
docker-compose up
```

The QLoRA training uses deepspeed package, and the deepspeed configuration file in "./antrea/ai/vicuna/playground". You can change the configuration by creating new configuration files and sending it into the "--deepspeed" argument in "docker-compose.yml".

If you're suffering with out-of-memory error with your GPU, you may try setting your MODEL_MAX_LENGTH and LORA_R lower. If you want to fully utilize your GPU memory you can also increase them or increase the PER_DEVICE_TRAIN_BATCH_SIZE. You can also set some environment variables such as PYTORCH_CUDA_ALLOC_CONF by different sizes of "max_split_size_mb" for GPU memory management. For more information, please refer to [the official pytorch docs of cuda](https://pytorch.org/docs/stable/notes/cuda.html#environment-variables).

After traininig, your trained QLoRA adapter models will be saved in MODEL_OUTPUT folder(mapped back to host path), for further testing. For loading trained adapters with the model, please refer to [Huggingface PeftModel.from_pretrained()](https://huggingface.co/docs/peft/en/package_reference/peft_model#peft.PeftModel).

> ## FAQ

Reviewer: ditto

Author: Updated

> # Vicuna Model Deployment and Training Documentation

Reviewer: Consider specifying the version of the model in the title to indicate the document's relevancy.

Author: updated

## Introduction

[Vicuna model](https://lmsys.org/blog/2023-03-30-vicuna/) is an open-source chatbot model trained by fine-tuning LLaMA on user-shared conversations collected from ShareGPT. It has superior performance compared to LLaMA and Alpaca, even approaching the performance of ChatGPT, evaluted by GPT-4 and with comparably low training costs. The code and model parameters [in open access with github](https://github.com/lm-sys/FastChat).

This document will focus on deploying and fine-tuning Vicuna model. We re-encapsulate the FastChat API for Vicuna, make it easier for personal use.

> Before use, you should have **GPU** in your system, with **enough GPU memory****(14GB for Vicuna-7B and 28GB for Vicuna-13B)**. Also, make sure that **docker, docker-compose, jq** are installed in your Linux system. 

Reviewer: Ensure that the statement "enough GPU memory (14GB for Vicuna-7B and 28GB for Vicuna-13B)" is always up-to-date. You can provide a reference to where users can verify this information.

Author: updated

> ## Deployment

Reviewer: Do we need to install Nvidia-docker before deployment? If so please add more details.

Author: updated

For Deployment of Vicuna, you should first change the directory:

```plain
cd ./antrea/ai/vicuna/deploy
```
We use Dockerfile to build docker container, you should have root privilege to run docker commands.
### 1.Build container with Vicuna

The environment variables should be set for build Vicuna container, so you should first set the environment variables MODEL_PATH and MODEL_NAMES in "1_build.sh". The MODEL_PATH should be accurate, and the MODEL_NAMES can be randomly named by your preference.

```plain

> export MODEL_PATH=<your_vicuna_model_path>

Reviewer: ```suggestion
export MODEL_PATH=<vicuna_model_path>
```

Author: updated

> export MODEL_NAMES=<random_name> 

Reviewer: ```suggestion
export MODEL_NAMES=<model_name> 
```

Author: updated

```
Then run the command:
```plain

> sudo sh 1_build.sh

Reviewer: Placing a number at the beginning of the script name does not adhere to the naming convention. Please remove it.

Author: updated

```
Since the docker-compose command in "1_build.sh", "sudo" is needed. This Shell scripts will first set the environment variables and then run the docker-compose to build container according to "docker-compose.yml" file, it needs 1-2 minutes or so. After container created successfully, you can send request to the url: [http://0.0.0.0:8000/v1/chat/completions](http://0.0.0.0:8000/v1/chat/completions)
### 2.Check if the model can repsonse

```plain
sh 2_call.sh
```
It will send POST request to call the model, and return the model reponses. If the responses are like this, the model has been deployed successfully. 
```plain
{"id":"chatcmpl-UFWP9BqDg3s7DDfc5eXLS3","object":"chat.completion","created":1701329914,"model":"vicuna","choices":[{"index":0,"message":{"role":"assistant","content":"Sure! Here are three numbers between 1 and 10:\n\n1. 5\n2. 8\n3. 9"},"finish_reason":"stop"}],"usage":{"prompt_tokens":51,"total_tokens":81,"completion_tokens":30}} 
```

> If the container has not been built, it may return that and you need to wait for a while. If it returns like this all the time, there may be something wrong.

Reviewer: ```suggestion
If the container has not been built, it may return this message, and you should wait for a while. If it consistently returns this message, there might be an issue:
```

Author: updated

```plain
{"object":"error","message":"Only  allowed now, your model vicuna","code":40301} 
```
### 3.Interactive with the model

```plain
sh 3_interactive.sh
```

> You can chat with the deployed Vicuna model by running this Shell script. You can ask any questions to the model interactively, and press Ctrl+Z to quit. The json processing tool **jq** is needed.

Reviewer: ```suggestion
You can chat with the deployed Vicuna model by running this Shell script. Feel free to ask any questions interactively to the model. To exit, simply press Ctrl+Z. Please note that the JSON processing tool **jq** is required for this operation.
```

Author: updated

# Vicuna Model Deployment and Training Documentation
## Introduction

[Vicuna model](https://lmsys.org/blog/2023-03-30-vicuna/) is an open-source chatbot model trained by fine-tuning LLaMA on user-shared conversations collected from ShareGPT. It has superior performance compared to LLaMA and Alpaca, even approaching the performance of ChatGPT, evaluted by GPT-4 and with comparably low training costs. The code and model parameters [in open access with github](https://github.com/lm-sys/FastChat).

This document will focus on deploying and fine-tuning Vicuna model. We re-encapsulate the FastChat API for Vicuna, make it easier for personal use. Currently, it works well with the lastest version of vicuna-7b-v1.5-16k and FastChat(v0.2.33). Please check the [version compatibility between Vicuna and FastChat](https://github.com/lm-sys/FastChat/blob/main/docs/vicuna_weights_version.md) if there's any issues of version.

> Before use, you should have **GPU** in your system, with **enough GPU memory** [**(14GB for Vicuna-7B and 28GB for Vicuna-13B)**](https://github.com/lm-sys/FastChat#single-gpu). Also, make sure that **docker, docker-compose, nvidia-docker, jq** are installed in your Linux system.

Reviewer: @zgy0817 I have a question here: Do we have to preinstall docker for nvidia-docker? Is it possible to directly install nvidia-docker without docker installed?

Author: Docker installed is a prerequisite for installing nvidia-docker, according to this link:
https://blog.csdn.net/qq_41776453/article/details/129794608

> # Vicuna Model Deployment and Training Documentation

Reviewer: Please add TOC(table of contents) to all docs, you can refer to this doc: https://github.com/antrea-io/antrea/blob/main/docs/windows.md#table-of-contents

Author: Updated
